src/strategies/AppserviceJoinRoomStrategy.ts

```diff
@@ -8,20 +8,24 @@ export class AppserviceJoinRoomStrategy implements IJoinRoomStrategy {
 
     public async joinRoom(roomIdOrAlias: string, userId: string, apiCall: (roomIdOrAlias: string) => Promise<string>): Promise<string> {
         try {
-            await apiCall(roomIdOrAlias);
+            return await apiCall(roomIdOrAlias);
         } catch (err) {
             LogService.error("AppserviceJoinRoomStrategy", err);
             if (userId !== this.appservice.botUserId) {
                 const client = this.appservice.botIntent.underlyingClient;
                 const roomId = await client.resolveRoom(roomIdOrAlias);
-                return client.inviteUser(userId, roomId).then(() => {
+                try {
+                    await client.inviteUser(userId, roomId);
+                } catch (err) {
                     if (this.underlyingStrategy) return this.underlyingStrategy.joinRoom(roomId, userId, apiCall);
```

Comment:
shouldn't this then be for the bot user ID? (with an await, and also keeping the `else` for joining the bot user)

Reply:
this catch only triggers if the appservice fails to invite the user. Right after the catch block we still do the if...else with joining the bot user.

Reply:
we don't though, because the userId never changes. If the invite fails, it is likely because the bridge isn't in the room, so we should join and then retry the invite. 

I think... this part of the library is a bit of a mess :(

Reply:
> If the invite fails, it is likely because the bridge isn't in the room, so we should join and then retry the invite.

or we are trying to join the ghost to a room where the appservice bridge isn't in. The appservice bridge bot doesn't need to be in a room to receive messages from synapse, only one of the ghosts is enough for that.
As a result for that, for direct puppetting 1:1 rooms it is nicest to not have the AS bot in the room but still the ghost.
To fix that, soru wanted to write another joining strategy to determine with which non-AS-bot to invite the ghost, but without this PR it just fails as, if the AS bot fails to invite, it never called back the next underlying strategy

Reply:
right, the puppeting use case would need its own join strategy. This one is intended for the default kind of bridging where a bridge bot should always be in the room and have high power levels.

Reply:
> right, the puppeting use case would need its own join strategy

The appservice class automatically adds the appservice join strategy, and additional strategies are chained to it, so without this PR you, well, can't do that at all. As ghosts are just appservice intents created via the appservice.

The intended usecase doesn't stop existing, the usecases for this join strategy are merely extended with no real downsides whatsoever

Reply:
Perhaps this alternative, identical way to write this section helps you understand what is going on:

```
return client.inviteUser(userId, roomId).then(() => {
    if (this.underlyingStrategy) return this.underlyingStrategy.joinRoom(roomId, userId, apiCall);
    else return apiCall(roomId);
}).catch((err) => {
    if (this.underlyingStrategy) return this.underlyingStrategy.joinRoom(roomId, userId, apiCall);
    throw err;
});
```

Reply:
hmm, I'm not sure I see how it can possibly support both use cases (ignoring the fact that the appservice stuff forces the use of it). I'll take a closer look later, but for now I think tests might be good to prove that both cases work.

Reply:
> hmm, I'm not sure I see how it can possibly support both use cases

perhaps have an optional flag on the underlyingStrategy and check for that?

Reply:
I think this is fine actually. I've finally just walked through the code and it will be fine.